projects/js-packages/publicize-components/src/hooks/use-schedule-post/index.tsx

```diff
@@ -0,0 +1,108 @@
+import { siteHasFeature } from '@automattic/jetpack-script-data';
+import { dispatch, useDispatch, useSelect } from '@wordpress/data';
+import { store as editorStore } from '@wordpress/editor';
+import { useCallback, useState } from '@wordpress/element';
+import { __ } from '@wordpress/i18n';
+import { store as noticesStore } from '@wordpress/notices';
+import { store as socialStore } from '../../social-store';
+import { features } from '../../utils';
+import useSocialMediaMessage from '../use-social-media-message';
+
+type SchedulePostOptions = {
+	/** The connection IDs to share to. */
+	connectionIds: number[];
+	/** The message to share. */
+	message?: string;
+	/** The timestamp to schedule the share for. */
+	timestamp: number;
+};
+
+/**
+ * Hook to schedule a post for sharing to social media connections.
+ *
+ * @return {object} Object containing schedule functionality and state.
+ */
+export function useSchedulePost() {
+	const [ isScheduling, setIsScheduling ] = useState( false );
+
+	const { postId, isAutosaveablePost, isDirtyPost } = useSelect( select => {
+		const editorSelector = select( editorStore );
+
+		return {
+			postId: editorSelector.getCurrentPostId(),
+			isAutosaveablePost: editorSelector.isEditedPostAutosaveable(),
+			isDirtyPost: editorSelector.isEditedPostDirty(),
+		};
+	}, [] );
+
+	const { createScheduledShare } = useDispatch( socialStore );
+	const { createSuccessNotice, createErrorNotice } = useDispatch( noticesStore );
+	const { message } = useSocialMediaMessage();
+	const hasMediaFeatures =
+		siteHasFeature( features.IMAGE_GENERATOR ) || siteHasFeature( features.ENHANCED_PUBLISHING );
+
+	const savePost = dispatch( editorStore ).savePost;
+
+	const schedulePost = useCallback(
+		async ( { connectionIds, timestamp }: SchedulePostOptions ) => {
+			if ( ! connectionIds.length || ! timestamp ) {
+				return false;
+			}
+
+			setIsScheduling( true );
+
+			/**
+			 * The share endpoint only gets the custom message as a parameter, the attached media and
+			 * SIG is saved to the post meta and will be read on wpcom. Because of that we need to save
+			 * the post before sharing it, if it has the media features to make sure we use the latest data.
+			 */
+			if ( isDirtyPost && isAutosaveablePost && hasMediaFeatures ) {
+				await savePost();
+			}
+
+			try {
+				// Process each connection separately, one at a time
+				for ( const connectionId of connectionIds ) {
+					await createScheduledShare( {
+						post_id: postId,
+						connection_id: connectionId,
+						message,
+						timestamp,
+					} );
+				}
+
+				createSuccessNotice(
+					__( 'Social post scheduled successfully.', 'jetpack-publicize-components' ),
+					{ type: 'snackbar' }
+				);
+
+				setIsScheduling( false );
+				return true;
+			} catch {
```

Comment:
I don't think `saveEntityRecord` throws an error if the request fails. We need to check this. You may need to return the result of `saveEntityRecord` call in `createScheduledShare` which is a boolean, implying whether the request was a success or not, IIRC.

Reply:
Correction: It seems like `saveEntityRecord` [does throw if an error occurs](https://github.com/WordPress/gutenberg/blob/9f77c96c2a474d1f7f90589a8f6965644be9a5a4/packages/core-data/src/actions.js#L690-L705).

Reply:
My suggestion would be wheter we should move this logic to the `createScheduledShare` action itself to show the notifications like we do for the delete action.

https://github.com/Automattic/jetpack/blob/121e9f80de3b82194e19419e7a69b8747ed7d34e/projects/js-packages/publicize-components/src/social-store/actions/scheduled-shares.ts#L18-L33

Reply:
I followed the pattern from use-share-post, but I can also move it to the action if you think that's better.

Reply:
Yeah, I think that will lift the burden off of the components/hooks.

Reply:
If a promise rejects then it should throw an exception, but saveEntityRecord looks to swallow it unless you pass the `throwOnError` option https://github.com/WordPress/gutenberg/blob/trunk/packages/core-data/src/actions.js#L502.

Perhaps we can have a batch action, that returns `Promise.all` and we can `await` here?

Reply:
We can return that `success` variable from the `createScheduledShare` action and it will be reflected in the `Promise.all()` call.

So, we can remove the `try...catch` and use this

```js
const result = await Promise.all(...);

return result.every( Boolean );
```